lib/nbp/push_connector.rb

```diff
@@ -0,0 +1,110 @@
+# frozen_string_literal: true
+
+module Nbp
+  class SettingsError < StandardError; end
+  class ConnectorError < StandardError; end
+
+  class PushConnector
+    include Singleton
+
+    def initialize
+      super
+      create_source! unless source_exists?
+    end
+
+    def push_lom!(xml)
+      response = api_conn.put("/push-connector/api/lom-v2/#{settings.source.slug}") do |req|
+        req.body = {metadata: xml}.to_json
+      end
+      raise_connector_error('Could not push task LOM', response) unless success_status?(response.status)
+    end
+
+    def delete_task!(task_uuid)
+      response = api_conn.delete("/push-connector/api/course/#{settings.source.slug}/#{task_uuid}")
+      raise_connector_error('Could delete task', response) unless success_status?(response.status)
+    end
+
+    def source_exists?
+      response = api_conn.get("/datenraum/api/core/sources/slug/#{settings.source.slug}")
+      if response.status == 200
+        true
+      elsif response.status == 404
+        false
+      else
+        raise_connector_error('Could not determine if source exists', response)
+      end
+    end
+
+    def create_source!
+      response = api_conn.post('/datenraum/api/core/sources') do |req|
+        req.body = settings.source.to_json
+      end
+      raise_connector_error('Failed to create source', response) unless success_status?(response.status)
+    end
+
+    def token
+      if @token.present? && @token_expiration > 10.seconds.from_now
+        @token
+      else
+        update_token
+      end
+    end
+
+    def update_token
+      response = Faraday.post(settings.token_path, auth)
+      result = JSON.parse(response.body)
+
+      if success_status?(response.status)
+        @token_expiration = Time.zone.now + result['expires_in']
+        @token = result['access_token']
+      else
+        raise_connector_error('Failed to get fresh access token', response)
+      end
+    end
+
+    def auth
+      {
+        grant_type: 'client_credentials',
+        client_id: settings.client_id,
+        client_secret: settings.client_secret,
+      }
+    end
+
+    def api_conn
+      @api_conn ||= Faraday.new(url: settings.api_host, headers:)
```

Comment:
Since the `api_conn` is memorized, it won't renew the `token`. Hence, five minutes after the first request, any further requests will fail with the current implementation.